src/driver/m3_subset_test_point_generator.py

```diff
@@ -0,0 +1,118 @@
+    sample subsets for testing.
+    """
+
+    def __init__(
+        self,
+        dataset: DataSet,
+        campaign_spend_fractions: np.ndarray,
+        max_num_points: int = 2000,
+        rng: np.random.Generator = np.random.default_rng(0),
+    ):
+        """Constructs an M3SubsetTestPointGenerator.
+
+        Args:
+            dataset:  The DataSet for which test points are to be generated.
+            campaign_spend_fractions:  The campaign_spend / inventory_spend
+                at each publisher, where each inventory_spend is given by
+                dataset._max_spends.
+            max_num_points:  If (2^p - 1) - (2p + 1) > `max_num_points`, then
+                randomly sample `max_num_points` subsets from the
+                (2^p - 1) - (2p + 1) subsets.
+            rng:  A numpy Generator object that is used to seed the generation
+                of random test points.
+        """
+        super().__init__(dataset)
+        self._rng = rng
+        self.max_num_points = max_num_points
+        self._campaign_spends = self._max_spends * campaign_spend_fractions
+
+    @staticmethod
+    def subset_layer(p: int, q: int) -> List[np.ndarray]:
+        """Returns a list of subset indicators for a layer of subsets.
+
+        For example, when p = 3, and q = 2, returns
+        [[1, 1, 0],  # indicating subset {A, B}
+         [1, 0, 1],  # indicating subset {A, C}
+         [0, 0, 1]]   # indicating subset {A, C}
+        In general, a subset indicator is a binary vector v where v[i] = 1
+        if and only if pub i is in the subset.
+        A layer of subsets means the collection of subsets with the same
+        cardinality.
+
+        Args:
+            p: Number of all publishers.
+            q: Number of publishers in the subsets.
+
+        Returns:
+            A length <p choose q> list of length <p> arrays.
+            Each array is a different binary vector that sums up to 1.
+        """
+
+        def one_direction(indices: List) -> np.ndarray:
+            x = np.zeros(p)
+            x[indices] = 1
+            return x
+
+        return [one_direction(list(indices)) for indices in combinations(range(p), q)]
+
+    def test_points(self) -> Iterable[List[float]]:
+        """Returns a generator for generating a list of test points.
+
+        Returns:
+            An iterable of spend vectors representing locations where
+            the true reach surface is to be compared to the modeled reach
+            surface.
+        """
+        p = self._npublishers
+        if p < 4:
+            return
+        if (2 ** p - 1) - (2 * p + 1) > self.max_num_points:
+            n = 0
+            while n < self.max_num_points:
+                subset_indicator = np.random.choice([0, 1], p)
+                num_involved_subsets = np.count_nonzero(subset_indicator)
+                if 1 < num_involved_subsets < p - 1:
+                    yield list(self._campaign_spends * subset_indicator)
```

Comment:
How do you ensure that each of the test points generated by this method is distinct?

Reply:
It's not ensured.  Some thoughts
1. It may be fine to allow the same test points?
2. If we want to ensure they are distinct, I can change the code using, say, `set`.  This adds computational complexity though.
PLMK what you think!

Reply:
Sure.  Just added a naive way to make them distinct, ptal.